File: src/python_general_pkg/python_general_pkg/simple_server_node.py
import rclpy
from rclpy.node import Node
from example_interfaces.srv import AddTwoInts


class SimpleServerNode(Node):
    def __init__(self, node_name):
        super().__init__(node_name)
        self._service = self.create_service(
            AddTwoInts, 
            "/python_simple_service", 
            self._callback, 
            )
        self.get_logger().info(f"服务节点 {node_name} 已启动!")

    def _callback(self, request, response):
        a = request.a
        b = request.b
        response.sum = a + b

        self.get_logger().info(f"收到请求: a={request.a}, b={request.b}")
        self.get_logger().info(f"返回响应: sum={response.sum}")
        return response


def main():
    rclpy.init()
    node = SimpleServerNode("python_simple_server")
    rclpy.spin(node)
    rclpy.shutdown()

File: src/cpp_general_pkg/src/simple_server_node.cpp
#include <string>

#include <rclcpp/rclcpp.hpp>
#include <example_interfaces/srv/add_two_ints.hpp>

using namespace std;
using std::placeholders::_1;
using std::placeholders::_2;


class SimpleServerNode: public rclcpp::Node {

public:
    SimpleServerNode(const string& node_name)
    : Node(node_name)
    {
        service_ = this->create_service<example_interfaces::srv::AddTwoInts>(
            "/cpp_simple_service",
            std::bind(&SimpleServerNode::callback, this, _1, _2));

        RCLCPP_INFO(this->get_logger(), "服务节点 cpp_simple_server 已启动!");
    }

private:
    rclcpp::Service<example_interfaces::srv::AddTwoInts>::SharedPtr service_;

    void callback(
        const std::shared_ptr<example_interfaces::srv::AddTwoInts::Request> request,
        std::shared_ptr<example_interfaces::srv::AddTwoInts::Response> response)
    {
        response->sum = request->a + request->b;

        RCLCPP_INFO(this->get_logger(), "收到请求: a=%ld, b=%ld", request->a, request->b);
        RCLCPP_INFO(this->get_logger(), "返回响应: sum=%ld", response->sum);
    }
};


int main(int argc, char** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<SimpleServerNode>("cpp_simple_server");
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
